### PR TITLE
DuckPlayer Launch Experiment for iOS

### PR DESCRIPTION
**Pull request by afterxleep**
Task/Issue URL: https://app.asana.com/0/1204099484721401/1208091265812750/f
Tech Design URL:  https://app.asana.com/0/1204099484721401/1207989846862252/f

**Description**:
We're running a launch experiment to try measure retention impact with DuckPlayer launch.  This is a pixel experiment, consisting in a few pixels sent in these scenarios:
- Search Performed
- Youtube Page visited

**Steps to test this PR**:
Code includes tests covering the correct pixel firing on different dates/weeks, etc.

### Setup
- [x] Make yourself Internal user
- [x] Drag content-scope-scripts to project from branch `pr-releases/release-993`
- [x] Add custom privacy config from `https://www.jsonblob.com/api/1276215786342309888`
- [x] Go to Settings and confirm 'Duck Player' is NOT visible
- [x] Filter your console messages with `duckplayer.experiment` (No category)

### Control Group
- [x] Modify [this](https://github.com/duckduckgo/iOS/pull/3328/files#diff-b024d67f44c561d35d8f144721f428379a29e52861b543c3902d668ce94401dcR142), so it always returns `control` (`let cohort: Cohort = .control`)
- [x] Build & Run
- [x] Go to Youtube.com and search for a 'metallica'
- [x] Tap the first video. 
- [x] Check your console and confirm the following two pixels were fired
- [x] Confirm variant is `control`, state is `disabled` and date matches today's date
- [x] Confirm that `youtube.page` pixel referer is `youtube`

**Example Pixels:**

```
Pixel fired duckplayer.experiment.cohort.assign ["enrollment": "20240905", "variant": "control"]
Pixel fired duckplayer.experiment.youtube.page.view ["variant": "control", "day": "0", "referrer": "youtube", "state": "alwaysAsk", "enrollment": "20240905"]
```

- [x] Refresh the page multiple times and confirm the pixel is not fired multiple times
- [x] Go back and select another video
- [x] Confirm the `duckplayer.experiment.youtube.page.view` pixel is fired again
- [x] Search for `black sabbath`using the URL Bar
- [x] Confirm the following 2 pixels are fired in the console

**Example pixels**
```
Pixel fired duckplayer.experiment.search ["enrollment": "20240905", "variant": "control", "day": "0"]
Pixel fired duckplayer.experiment.daily.search ["enrollment": "20240905", "variant": "control", "day": "0"]
```

- [x] Run another search for `black sabbath war pigs`  and confirm only one pixel is fired in the console

`Pixel fired duckplayer.experiment.search ["day": "0", "variant": "control", "enrollment": "20240905"]`

- [x] Tap the first video from the results 
- [x] Confirm the following pixel is fired in the console, and referrer is `serp`

**Example Pixel**
```
Pixel fired duckplayer.experiment.youtube.page.view ["enrollment": "20240905", "state": "disabled", "variant": "control", "day": "0", "referrer": "serp"]
```

- [x] Go to google.com and search for `number of the beast` 
- [x] Tap the first link to Youtube video (make sure it opens in Youtube)
- [x] Confirm the `experiment.search` pixel is fired and referrer is `other`

**Example Pixel**
```
Pixel fired duckplayer.experiment.youtube.page.view ["variant": "control", "day": "0", "state": "disabled", "referrer": "other", "enrollment": "20240905"]
```

### Experiment Group
- [x] Modify [this](https://github.com/duckduckgo/iOS/pull/3328/files#diff-b024d67f44c561d35d8f144721f428379a29e52861b543c3902d668ce94401dcR142), so it always returns `experiment` (`let cohort: Cohort = .experiment`)
- [x] Go to Settings > Debug > All Debug Options > Reset DuckPlayer Experiment
- [x] Close all tabs
- [x] Go to Settings, and confirm DuckPlayer is NOT visible (This is expected)
- [x] Go to Youtube.com and search for 'creedence'
- [x] Tap the first video. 
- [x] Confirm DuckPlayer Overlay is visible and both the enrollment pixel and video view pixels are there
- [x] Confirm that the referrer in the view pixel is 'alwaysAsk' or whatever setting you had for DuckPlayer before
- [x] Tap Turn On Duck Player
- [x] Confirm DuckPlayer shows up and No additional pixel is fired
- [x] Add page to favorites
- [x] Go back and navigate to another video
- [x] Close all tabs and open the favorite you've added before..  Confirm a view pixel is fired again
- [x] Go to Settings, and confirm DuckPlayer IS now visible

**Example Pixel for experimen view**
```
Pixel fired duckplayer.experiment.youtube.page.view ["variant": "control", "day": "0", "state": "alwaysAsk", "referrer": "other", "enrollment": "20240905"]
```

### Other
- [ ] Smoke test DuckPlayer (Disable, enable, watch videos, etc)

### CODE REVIEW

File: DuckDuckGo/DuckPlayer/DuckPlayer.swift [hunk 1]
```diff
@@ -110,8 +123,9 @@ final class DuckPlayer: DuckPlayerProtocol {
         static let featureNameKey = "featureName"
     }
     
-    private(set) var settings: DuckPlayerSettingsProtocol
```
Comment (afterxleep) [inline DuckDuckGo/DuckPlayer/DuckPlayer.swift:113]: Renamed the protocol to `DuckPlayerSettings`, which makes more sense

File: DuckDuckGo/DuckPlayer/DuckPlayer.swift [hunk 2]
```diff
@@ -216,8 +232,11 @@ final class DuckPlayer: DuckPlayerProtocol {
     }
 
     private func encodeUserValues() -> UserValues {
-        UserValues(
-            duckPlayerMode: settings.mode,
+        let experiment = DuckPlayerLaunchExperiment()
+        let isEnabled = experiment.isEnrolled && experiment.isExperimentCohort && featureFlagger.isFeatureOn(.duckPlayer)
```
Comment (afterxleep) [inline DuckDuckGo/DuckPlayer/DuckPlayer.swift]: These values are the ones passed to the FE, so account for experiment group and feature flag

File: DuckDuckGo/DuckPlayer/DuckPlayerLaunchExperiment.swift [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        self.dateProvider = dateProvider
+    }
+    
+    private var dates: (day: Int, week: Int)? {
+        guard isEnrolled,
+              let enrollmentDate = enrollmentDate else { return nil }
+        let currentDate = dateProvider.currentDate
+        let calendar = Calendar.current
+        let dayDifference = calendar.dateComponents([.day], from: enrollmentDate, to: currentDate).day ?? 0
+        let weekDifference = (dayDifference / 7)  + 1
```
Comment (afterxleep) [inline DuckDuckGo/DuckPlayer/DuckPlayerLaunchExperiment.swift:125]: Week "0" is not a thing.  We expect it to be 1 or 2, but Day "0" is OK

File: DuckDuckGo/DuckPlayer/DuckPlayerNavigationHandler.swift [hunk 4]
```diff
@@ -397,17 +456,28 @@ extension DuckPlayerNavigationHandler: DuckNavigationHandling {
...
+            duckPlayerMode == .enabled || duckPlayerMode == .alwaysAsk {
             Logger.duckPlayer.debug("DP: Handling Initial Load of a video for \(url.absoluteString)")
             handleReload(webView: webView)
         }
         
     }
-
+    
+    // Handle custom events
+    func handleEvent(event: DuckPlayerNavigationEvent, url: URL?, navigationAction: WKNavigationAction?) {
```
Comment (afterxleep) [inline DuckDuckGo/DuckPlayer/DuckPlayerNavigationHandler.swift:489]: The idea here is to handle 'Custom' events that are unrelated to DuckPlayer as a feature but not with presentation.-

File: DuckDuckGo/MainViewController+Segues.swift [hunk 5]
```diff
@@ -175,7 +175,6 @@ extension MainViewController {
 
     private func fireBrokenSiteReportShown() {
         let parameters = [
-            PrivacyDashboardEvents.Parameters.variant: PixelExperiment.privacyDashboardVariant.rawValue,
```
Comment (afterxleep) [inline DuckDuckGo/MainViewController+Segues.swift:178]: There was a bunch of old PixelExperiment stuff that was not cleaned up before 🤷

**Review by Bunn — APPROVED** [review]
Code is OK for experiment.

**Pull request merged**